src/i2cint.rs

```diff
@@ -0,0 +1,386 @@
+    /// * `I2cError::DeviceBusy` if the device is already busy.
+    fn _write(&mut self, addr: u8, auto_stop: bool, bytes: &'static [u8]) -> Result<(), I2cError> {
+        if self.is_busy() {
+            self.last_error = Some(I2cError::DeviceBusy);
+            return Err(I2cError::DeviceBusy);
+        }
+        self.tx_ind = 0;
+        self.tx_buf = Some(bytes);
+        self.write_start(addr, bytes.len() as u8, auto_stop);
+        Ok(())
+    }
+
+    /// Start a write sequence on i2c channel.
+    ///
+    /// # Arguments
+    /// * `addr` - Destination address.
+    /// * `n_bytes` - number of bytes which will be sent.
+    /// * `auto_stop` - i2c autostop enabled.
+    fn write_start(&mut self, addr: u8, n_bytes: u8, auto_stop: bool) {
+        self.current_write_addr = Some(addr);
+        self.dev.i2c.cr2.modify(|_, w| {
+            w.sadd()
+                .bits(u16::from(addr << 1))
+                .rd_wrn()
+                .write()
+                .nbytes()
+                .bits(n_bytes)
+                .start()
+                .start()
+                .autoend()
+                .bit(auto_stop)
+        });
+        self.state = State::TxStart;
+    }
+
+    /// Reads from i2c interface. Supports only read < 256 bytes.
+    ///
+    /// # Arguments
+    /// * `addr` - Destination address.
+    /// * `len` - number of bytes to read.
+    pub fn read(&mut self, addr: u8, len: usize) {
+        self.dev.i2c.cr2.modify(|_, w| {
+            w.sadd()
+                .bits(u16::from(addr << 1))
+                .rd_wrn()
+                .read()
+                .nbytes()
+                .bits(len as u8)
+                .start()
+                .start()
+                .autoend()
+                .automatic()
+        });
+        self.state = State::RxStart;
+    }
+
+    /// Write bytes through i2c. Supports only write and read < 256 bytes.
+    ///
+    /// # Arguments
+    /// * `addr` - Destination address.
+    /// * `bytes` - Bytes to send.
+    /// * `recv_len` - Number of bytes to receive.
+    ///
+    /// # Errors
+    /// * `I2cError::DeviceBusy` if the device is already busy.
+    pub fn write_read(
+        &mut self,
+        addr: u8,
+        bytes: &'static [u8],
+        recv_len: usize,
+    ) -> Result<(), I2cError> {
+        self.recv = Some((addr, recv_len));
+        self._write(addr, false, bytes)
+    }
+
+    /// Send a stop.
+    ///
+    /// # Arguments
+    /// * `addr` - Destination address.
+    ///
+    /// # Errors
+    /// * `I2cError::DeviceBusy` if the device is already busy.
+    pub fn stop(&mut self, addr: u8) {
+        self.dev
+            .i2c
+            .cr2
+            .modify(|_, w| w.sadd().bits(u16::from(addr << 1)).stop().stop());
+    }
+
+    /// This function must be called when there is an interruption on i2c device.
+    /// It will compute the current state based on ISR register and execute work based on the state.
+    pub fn interrupt(&mut self) {
```

Comment:
`interrupt` sound like, this function would return the type of the interrupt, which got triggered rather than reaction on the interrupt. I'd look for a more descriptive name. Maybe `react_on_interrupt()`? Better suggestions are welcome.